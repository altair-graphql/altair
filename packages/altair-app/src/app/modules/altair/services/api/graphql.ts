import { gql } from 'apollo-angular';
import { Injectable } from '@angular/core';
import * as Apollo from 'apollo-angular';
import * as ApolloCore from '@apollo/client/core';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type CreateRequestCollectionInput = {
  name: Scalars['String'];
  workspaceId?: InputMaybe<Scalars['String']>;
};

export type CreateRequestInput = {
  collectionId?: InputMaybe<Scalars['String']>;
  headers?: InputMaybe<Array<InputMaybe<HeaderInput>>>;
  name?: InputMaybe<Scalars['String']>;
  ownerId?: InputMaybe<Scalars['String']>;
  query?: InputMaybe<Scalars['String']>;
  url?: InputMaybe<Scalars['String']>;
  variables?: InputMaybe<Scalars['String']>;
};

export type CreateUserInput = {
  email: Scalars['String'];
  firstName: Scalars['String'];
  lastName?: InputMaybe<Scalars['String']>;
};

export type CreateWorkspaceInput = {
  name: Scalars['String'];
};

export type Header = {
  __typename?: 'Header';
  key: Scalars['String'];
  value: Scalars['String'];
};

export type HeaderInput = {
  key: Scalars['String'];
  value: Scalars['String'];
};

export type Mutation = {
  __typename?: 'Mutation';
  createRequest: Request;
  createRequestCollection: RequestCollection;
  createWorkspace: Workspace;
  getOrCreateUser?: Maybe<User>;
  removeRequest?: Maybe<Request>;
  removeRequestCollection?: Maybe<RequestCollection>;
  removeWorkspace?: Maybe<Workspace>;
  updateRequest: Request;
  updateRequestCollection: RequestCollection;
  updateWorkspace: Workspace;
};


export type MutationCreateRequestArgs = {
  createRequestInput: CreateRequestInput;
};


export type MutationCreateRequestCollectionArgs = {
  createRequestCollectionInput: CreateRequestCollectionInput;
};


export type MutationCreateWorkspaceArgs = {
  createWorkspaceInput: CreateWorkspaceInput;
};


export type MutationRemoveRequestArgs = {
  id: Scalars['Int'];
};


export type MutationRemoveRequestCollectionArgs = {
  id: Scalars['Int'];
};


export type MutationRemoveWorkspaceArgs = {
  id: Scalars['String'];
};


export type MutationUpdateRequestArgs = {
  updateRequestInput: UpdateRequestInput;
};


export type MutationUpdateRequestCollectionArgs = {
  updateRequestCollectionInput: UpdateRequestCollectionInput;
};


export type MutationUpdateWorkspaceArgs = {
  updateWorkspaceInput: UpdateWorkspaceInput;
};

export type Query = {
  __typename?: 'Query';
  getUserById?: Maybe<User>;
  profile?: Maybe<User>;
  request?: Maybe<Request>;
  requestCollection?: Maybe<RequestCollection>;
  requestCollections: Array<Maybe<RequestCollection>>;
  requests: Array<Maybe<Request>>;
  workspace?: Maybe<Workspace>;
  workspaces: Array<Maybe<Workspace>>;
};


export type QueryGetUserByIdArgs = {
  id: Scalars['String'];
};


export type QueryRequestArgs = {
  id: Scalars['Int'];
};


export type QueryRequestCollectionArgs = {
  id: Scalars['Int'];
};


export type QueryWorkspaceArgs = {
  id: Scalars['Int'];
};

export type Request = {
  __typename?: 'Request';
  collectionId?: Maybe<Scalars['String']>;
  headers?: Maybe<Array<Maybe<Header>>>;
  name?: Maybe<Scalars['String']>;
  ownerId?: Maybe<Scalars['String']>;
  query?: Maybe<Scalars['String']>;
  url?: Maybe<Scalars['String']>;
  variables?: Maybe<Scalars['String']>;
};

export type RequestCollection = {
  __typename?: 'RequestCollection';
  name?: Maybe<Scalars['String']>;
  workspaceId?: Maybe<Scalars['String']>;
};

export type UpdateRequestCollectionInput = {
  id: Scalars['Int'];
};

export type UpdateRequestInput = {
  id: Scalars['Int'];
};

export type UpdateWorkspaceInput = {
  id: Scalars['String'];
  name?: InputMaybe<Scalars['String']>;
};

export type User = {
  __typename?: 'User';
  email: Scalars['String'];
  firstName: Scalars['String'];
  id: Scalars['String'];
  lastName?: Maybe<Scalars['String']>;
};

export type Workspace = {
  __typename?: 'Workspace';
  id?: Maybe<Scalars['String']>;
  name?: Maybe<Scalars['String']>;
};

export type ProfileQueryVariables = Exact<{ [key: string]: never; }>;


export type ProfileQuery = { __typename?: 'Query', profile?: { __typename?: 'User', email: string, firstName: string, lastName?: string | null | undefined } | null | undefined };

export type GetOrCreateUserMutationVariables = Exact<{ [key: string]: never; }>;


export type GetOrCreateUserMutation = { __typename?: 'Mutation', getOrCreateUser?: { __typename?: 'User', email: string, firstName: string, lastName?: string | null | undefined } | null | undefined };

export const ProfileDocument = gql`
    query Profile {
  profile {
    email
    firstName
    lastName
  }
}
    `;

  @Injectable({
    providedIn: 'root'
  })
  export class ProfileGQL extends Apollo.Query<ProfileQuery, ProfileQueryVariables> {
    document = ProfileDocument;
    
    constructor(apollo: Apollo.Apollo) {
      super(apollo);
    }
  }
export const GetOrCreateUserDocument = gql`
    mutation getOrCreateUser {
  getOrCreateUser {
    email
    firstName
    lastName
  }
}
    `;

  @Injectable({
    providedIn: 'root'
  })
  export class GetOrCreateUserGQL extends Apollo.Mutation<GetOrCreateUserMutation, GetOrCreateUserMutationVariables> {
    document = GetOrCreateUserDocument;
    
    constructor(apollo: Apollo.Apollo) {
      super(apollo);
    }
  }

  type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

  interface WatchQueryOptionsAlone<V>
    extends Omit<ApolloCore.WatchQueryOptions<V>, 'query' | 'variables'> {}
    
  interface QueryOptionsAlone<V>
    extends Omit<ApolloCore.QueryOptions<V>, 'query' | 'variables'> {}
    
  interface MutationOptionsAlone<T, V>
    extends Omit<ApolloCore.MutationOptions<T, V>, 'mutation' | 'variables'> {}
    
  interface SubscriptionOptionsAlone<V>
    extends Omit<ApolloCore.SubscriptionOptions<V>, 'query' | 'variables'> {}

  @Injectable({ providedIn: 'root' })
  export class APIClient {
    constructor(
      private profileGql: ProfileGQL,
      private getOrCreateUserGql: GetOrCreateUserGQL
    ) {}
      
    profile(variables?: ProfileQueryVariables, options?: QueryOptionsAlone<ProfileQueryVariables>) {
      return this.profileGql.fetch(variables, options)
    }
    
    profileWatch(variables?: ProfileQueryVariables, options?: WatchQueryOptionsAlone<ProfileQueryVariables>) {
      return this.profileGql.watch(variables, options)
    }
    
    getOrCreateUser(variables?: GetOrCreateUserMutationVariables, options?: MutationOptionsAlone<GetOrCreateUserMutation, GetOrCreateUserMutationVariables>) {
      return this.getOrCreateUserGql.mutate(variables, options)
    }
  }