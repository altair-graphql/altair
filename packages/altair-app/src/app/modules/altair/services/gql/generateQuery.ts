import {
  ArgumentNode,
  astFromValue,
  DocumentNode,
  FragmentDefinitionNode,
  getNamedType,
  GraphQLBoolean,
  GraphQLEnumType,
  GraphQLField,
  GraphQLInputObjectType,
  GraphQLInputType,
  GraphQLInt,
  GraphQLInterfaceType,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString,
  GraphQLType,
  isInputObjectType,
  isLeafType,
  isListType,
  isNonNullType,
  isObjectType,
  isRequiredInputField,
  Kind,
  OperationDefinitionNode,
  OperationTypeNode,
  print,
  SelectionSetNode,
  ValueNode,
} from 'graphql';
import { Maybe } from 'graphql/jsutils/Maybe';
import { getTypeName } from './helpers';
import { prettify } from './prettifier';

interface GenerateMeta {
  hasArgs?: boolean;
}

export const generateQuery = async (
  schema: GraphQLSchema,
  name: string,
  parentType: string,
  { maxDepth = 2, tabSize = 2 } = {}
) => {
  const parentTypeObject = schema.getType(parentType);

  if (!(parentTypeObject instanceof GraphQLObjectType)) {
    return { generated: '', metas: [] };
  }
  const field = parentTypeObject.getFields()[name];

  if (!field) {
    return { generated: '', metas: [] };
  }

  const operationType = getOperationType(schema, parentTypeObject);

  const { node: selectionNode, metas } = buildSelectionNode(field, {
    maxDepth,
    currentDepth: 0,
  });
  const selectionSet = {
    kind: Kind.SELECTION_SET,
    // start building actual content here
    selections: [selectionNode],
  };

  const x: DocumentNode = {
    kind: Kind.DOCUMENT,
    definitions: [
      operationType
        ? buildOperationDefinitionWrapper(operationType, selectionSet)
        : buildFragmentDefinitionWrapper(parentTypeObject, selectionSet),
    ],
  };

  return {
    generated: await prettify(print(x), tabSize),
    metas,
  };
};

const getOperationType = (schema: GraphQLSchema, type: GraphQLObjectType) => {
  const typeToOperation = new Map<
    Maybe<GraphQLObjectType<any, any>>,
    OperationTypeNode
  >([
    [schema.getQueryType(), 'query'],
    [schema.getMutationType(), 'mutation'],
    [schema.getSubscriptionType(), 'subscription'],
  ]);

  return typeToOperation.get(type);
};

// operation definition would wrap a selection to return a query AST (for fields from root types)
const buildOperationDefinitionWrapper = (
  operationType: OperationTypeNode,
  selectionSet: SelectionSetNode
): OperationDefinitionNode => {
  return {
    kind: Kind.OPERATION_DEFINITION,
    operation: operationType,
    name: undefined,
    variableDefinitions: [],
    directives: [],
    selectionSet,
  };
};

// fragment definition would wrap a selection to return a fragment AST
const buildFragmentDefinitionWrapper = (
  parentType: GraphQLObjectType,
  selectionSet: SelectionSetNode
): FragmentDefinitionNode => {
  return {
    kind: Kind.FRAGMENT_DEFINITION,
    name: {
      kind: Kind.NAME,
      value: '____',
    },
    variableDefinitions: [],
    directives: [],
    typeCondition: {
      kind: Kind.NAMED_TYPE,
      name: {
        kind: Kind.NAME,
        value: parentType.name,
      },
    },
    selectionSet,
  };
};

export const buildSelectionSet = (
  type: GraphQLType | null,
  { maxDepth = 1, currentDepth = 0 } = {}
): { selectionSet: SelectionSetNode; metas: GenerateMeta[] } => {
  let setMetas: GenerateMeta[] = [];
  const selectionSet: SelectionSetNode = {
    kind: Kind.SELECTION_SET,
    selections: [],
  };

  if (!type) {
    return { selectionSet, metas: setMetas };
  }
  const namedType = getNamedType(type);

  if (
    !(
      namedType instanceof GraphQLObjectType ||
      namedType instanceof GraphQLInterfaceType
    )
  ) {
    return { selectionSet, metas: setMetas };
  }

  if (isLeafType(type)) {
    return { selectionSet, metas: setMetas };
  }

  if (currentDepth >= maxDepth) {
    return { selectionSet, metas: setMetas };
  }

  const fields = namedType && namedType.getFields();
  selectionSet.selections = Object.entries(fields).map(([fieldName, field]) => {
    const { node, metas } = buildSelectionNode(field, {
      maxDepth,
      currentDepth,
    });

    setMetas = [...setMetas, ...metas];

    return node;
  });

  return { selectionSet, metas: setMetas };
};

export const buildSelectionNode = (
  field: GraphQLField<any, any>,
  { maxDepth = 1, currentDepth = 0 } = {}
) => {
  const argumentsNodes: ArgumentNode[] = field.args.map((arg) => {
    return {
      kind: Kind.ARGUMENT,
      name: {
        kind: Kind.NAME,
        value: arg.name,
      },
      value: buildDefaultArgumentValueNode(arg.type, arg.defaultValue),
    };
  });

  const { selectionSet, metas } = buildSelectionSet(field.type, {
    maxDepth,
    currentDepth: currentDepth + 1,
  });

  const node = {
    kind: Kind.FIELD,
    alias: undefined,
    name: {
      kind: Kind.NAME,
      value: field.name,
    },
    arguments: argumentsNodes,
    directives: [],
    selectionSet,
  } as const;

  return { node, metas: [{ hasArgs: !!argumentsNodes.length }, ...metas] };
};

// Generate default values based on the GraphQL type
const buildDefaultArgumentValueNode = (
  argumentType: GraphQLInputType,
  defaultValue?: unknown,
  { currentDepth = 0, maxDepth = 2 } = {}
): ValueNode => {
  return (
    maybeBuildDefaultArgumentValueNode(argumentType, defaultValue, {
      currentDepth,
      maxDepth,
    }) ?? { kind: Kind.STRING, value: '_____' }
  );
};
const maybeBuildDefaultArgumentValueNode = (
  type: GraphQLInputType,
  defaultValue?: unknown,
  { currentDepth = 0, maxDepth = 2 } = {}
): ValueNode | undefined | null => {
  const defaultValueNode = astFromValue(defaultValue, type);
  if (defaultValueNode) {
    return defaultValueNode;
  }

  if (isNonNullType(type)) {
    type = type.ofType;
  }

  // handle lists
  if (isListType(type)) {
    const value: ValueNode = {
      kind: Kind.LIST,
      values: [
        buildDefaultArgumentValueNode(type.ofType, undefined, {
          currentDepth: currentDepth + 1,
          maxDepth,
        }),
      ],
    };
    return value;
  }

  // handle complex objects
  if (isObjectType(type) || isInputObjectType(type)) {
    if (currentDepth >= maxDepth) {
      return;
    }

    const fieldNodes = [];
    for (const field of Object.values(
      (type as GraphQLInputObjectType).getFields()
    )) {
      if (isRequiredInputField(field)) {
        const fieldValue = maybeBuildDefaultArgumentValueNode(
          field.type,
          undefined,
          { currentDepth: currentDepth + 1, maxDepth }
        );
        if (fieldValue) {
          fieldNodes.push({
            kind: Kind.OBJECT_FIELD,
            name: { kind: Kind.NAME, value: field.name },
            value: fieldValue,
          });
        }
      }
    }
    return { kind: Kind.OBJECT, fields: fieldNodes };
  }

  if (type instanceof GraphQLEnumType) {
    const defaultValue = type.getValues()[0];

    if (!defaultValue) {
      return;
    }

    return astFromValue(defaultValue.name, type);
  }

  const typeName = getTypeName(type);

  switch (typeName) {
    case 'String': {
      return astFromValue('string', GraphQLString);
    }
    case 'Boolean': {
      return astFromValue(true, GraphQLBoolean);
    }
    case 'Int': {
      return astFromValue(1, GraphQLInt);
    }
  }
};
