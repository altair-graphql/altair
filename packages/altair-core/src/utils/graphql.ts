import { DocumentNode, Kind, OperationDefinitionNode, parse } from 'graphql';

const getEmptyDocumentNode = (): DocumentNode => {
  return {
    definitions: [],
    kind: Kind.DOCUMENT,
  };
};
const parseQueryOrEmptyDocument = (query: string) => {
  if (!query) {
    return getEmptyDocumentNode();
  }

  try {
    return parse(query);
  } catch (err) {
    console.error('Error parsing query, returning empty instead', err);

    return getEmptyDocumentNode();
  }
};

export const getOperations = (query: string) => {
  const parsedQuery = parseQueryOrEmptyDocument(query);

  if (parsedQuery.definitions) {
    return parsedQuery.definitions.filter(
      (def): def is OperationDefinitionNode =>
        !!(def.kind === 'OperationDefinition' && def.name?.value)
    );
  }

  return [];
};
