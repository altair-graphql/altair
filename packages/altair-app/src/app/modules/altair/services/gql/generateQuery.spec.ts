import fs from 'fs';
import path from 'path';
import { describe, it, expect } from '@jest/globals';
import { buildSchema } from 'graphql';
import { generateQuery } from './generateQuery';

const getTestSchema = () => {
  const sdl = fs.readFileSync(
    path.resolve(__dirname, '../../../../../../fixtures/test.sdl'),
    'utf8'
  );
  return buildSchema(sdl);
};

describe('generateQuery', () => {
  describe('query', () => {
    it('returns empty string if parent type is invalid', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'GOTHouses', 'Invalid', {});
      expect(res.generated).toBe('');
    });
    it('returns empty string if field is invalid', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'invalid', 'Query', {});
      expect(res.generated).toBe('');
    });
    it('generates expected query', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'GOTHouses', 'Query', {});
      expect(res.generated).toMatchSnapshot();
    });
    it('generates expected query with specified max depth', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'GOTHouses', 'Query', {
        maxDepth: 5,
      });
      expect(res.generated).toMatchSnapshot();
    });
    it('generates expected query with specified tab size', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'GOTHouses', 'Query', {
        tabSize: 5,
      });
      expect(res.generated).toMatchSnapshot();
    });
    it('generates expected query with default string argument', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'GOTBooks', 'Query', {
        maxDepth: 2,
      });
      expect(res.generated).toMatchSnapshot();
    });
    it('generates expected query with default enum argument', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'files', 'Query', {
        maxDepth: 2,
      });
      expect(res.generated).toMatchSnapshot();
    });
    it('generates expected query with string argument', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'withString', 'Query', {
        maxDepth: 2,
      });
      expect(res.generated).toMatchSnapshot();
    });
    it('generates expected query with int argument', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'withInt', 'Query', {
        maxDepth: 2,
      });
      expect(res.generated).toMatchSnapshot();
    });
    it('generates expected query with boolean argument', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'withBoolean', 'Query', {
        maxDepth: 2,
      });
      expect(res.generated).toMatchSnapshot();
    });
    it('generates expected query with enum argument', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'withEnum', 'Query', {
        maxDepth: 2,
      });
      expect(res.generated).toMatchSnapshot();
    });
    it('generates expected query with list argument', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'withStringList', 'Query', {
        maxDepth: 2,
      });
      expect(res.generated).toMatchSnapshot();
    });
    it('generates expected query with object argument', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'withGOTCharacter', 'Query', {
        maxDepth: 2,
      });
      expect(res.generated).toMatchSnapshot();
    });
    it('generates expected query with multiple complex arguments', async () => {
      const schema = getTestSchema();
      const res = await generateQuery(schema, 'withMultipleListArguments', 'Query', {
        maxDepth: 2,
      });
      expect(res.generated).toMatchSnapshot();
    });
  });

  describe('fragment', () => {
    it('generates expected fragment', async () => {
      const schema = getTestSchema();

      const res = await generateQuery(schema, 'overlord', 'GOTHouse', {});
      expect(res.generated).toMatchSnapshot();
    });
    it('generates expected fragment with specified max depth', async () => {
      const schema = getTestSchema();

      const res = await generateQuery(schema, 'overlord', 'GOTHouse', {
        maxDepth: 5,
      });
      expect(res.generated).toMatchSnapshot();
    });
    it('generates expected fragment with specified tab size', async () => {
      const schema = getTestSchema();

      const res = await generateQuery(schema, 'overlord', 'GOTHouse', {
        tabSize: 5,
      });
      expect(res.generated).toMatchSnapshot();
    });
  });
});
