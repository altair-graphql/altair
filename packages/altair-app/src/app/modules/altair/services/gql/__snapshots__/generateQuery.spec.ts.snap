// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateQuery fragment generates expected fragment 1`] = `
"fragment ____ on GOTHouse {
  overlord {
    id
    url
    name
    region
    titles
    seats
    words
    overlord
    swornMembers
  }
}
"
`;

exports[`generateQuery fragment generates expected fragment with specified max depth 1`] = `
"fragment ____ on GOTHouse {
  overlord {
    id
    url
    name
    region
    titles
    seats
    words
    overlord {
      id
      url
      name
      region
      titles
      seats
      words
      overlord {
        id
        url
        name
        region
        titles
        seats
        words
        overlord {
          id
          url
          name
          region
          titles
          seats
          words
          overlord
          swornMembers
        }
        swornMembers {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
      }
      swornMembers {
        id
        book {
          id
          url
          name
          authors
          characters
          released
        }
        url
        name
        gender
        culture
        born
        died
        titles
        aliases
        father {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        mother {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        spouse {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        allegiances
        books
        playedBy
      }
    }
    swornMembers {
      id
      book {
        id
        url
        name
        authors
        characters {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        released
      }
      url
      name
      gender
      culture
      born
      died
      titles
      aliases
      father {
        id
        book {
          id
          url
          name
          authors
          characters
          released
        }
        url
        name
        gender
        culture
        born
        died
        titles
        aliases
        father {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        mother {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        spouse {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        allegiances
        books
        playedBy
      }
      mother {
        id
        book {
          id
          url
          name
          authors
          characters
          released
        }
        url
        name
        gender
        culture
        born
        died
        titles
        aliases
        father {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        mother {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        spouse {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        allegiances
        books
        playedBy
      }
      spouse {
        id
        book {
          id
          url
          name
          authors
          characters
          released
        }
        url
        name
        gender
        culture
        born
        died
        titles
        aliases
        father {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        mother {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        spouse {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        allegiances
        books
        playedBy
      }
      allegiances
      books
      playedBy
    }
  }
}
"
`;

exports[`generateQuery fragment generates expected fragment with specified tab size 1`] = `
"fragment ____ on GOTHouse {
     overlord {
          id
          url
          name
          region
          titles
          seats
          words
          overlord
          swornMembers
     }
}
"
`;

exports[`generateQuery query generates expected query 1`] = `
"{
  GOTHouses(name: "string") {
    id
    url
    name
    region
    titles
    seats
    words
    overlord
    swornMembers
  }
}
"
`;

exports[`generateQuery query generates expected query with boolean argument 1`] = `
"{
  withBoolean(isValid: true) {
    id
    url
    name
    authors
    characters
    released
  }
}
"
`;

exports[`generateQuery query generates expected query with default enum argument 1`] = `
"{
  files(filetype: PDF) {
    filename
    mimetype
    encoding
    filepath
  }
}
"
`;

exports[`generateQuery query generates expected query with default string argument 1`] = `
"{
  GOTBooks(name: "my-book") {
    id
    url
    name
    authors
    characters
    released
  }
}
"
`;

exports[`generateQuery query generates expected query with enum argument 1`] = `
"{
  withEnum(name: JPG) {
    id
    url
    name
    authors
    characters
    released
  }
}
"
`;

exports[`generateQuery query generates expected query with int argument 1`] = `
"{
  withInt(count: 1) {
    id
    url
    name
    authors
    characters
    released
  }
}
"
`;

exports[`generateQuery query generates expected query with list argument 1`] = `
"{
  withStringList(list: ["string"]) {
    id
    url
    name
    authors
    characters
    released
  }
}
"
`;

exports[`generateQuery query generates expected query with multiple complex arguments 1`] = `
"{
  withMultipleListArguments(
    list: [{ id: 1 }]
    list2: [true]
    character: { id: 1, book: { id: 1 } }
  ) {
    id
    url
    name
    authors
    characters
    released
  }
}
"
`;

exports[`generateQuery query generates expected query with object argument 1`] = `
"{
  withGOTCharacter(character: { id: 1, book: { id: 1 } }) {
    id
    url
    name
    region
    titles
    seats
    words
    overlord
    swornMembers
  }
}
"
`;

exports[`generateQuery query generates expected query with specified max depth 1`] = `
"{
  GOTHouses(name: "string") {
    id
    url
    name
    region
    titles
    seats
    words
    overlord {
      id
      url
      name
      region
      titles
      seats
      words
      overlord {
        id
        url
        name
        region
        titles
        seats
        words
        overlord {
          id
          url
          name
          region
          titles
          seats
          words
          overlord
          swornMembers
        }
        swornMembers {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
      }
      swornMembers {
        id
        book {
          id
          url
          name
          authors
          characters
          released
        }
        url
        name
        gender
        culture
        born
        died
        titles
        aliases
        father {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        mother {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        spouse {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        allegiances
        books
        playedBy
      }
    }
    swornMembers {
      id
      book {
        id
        url
        name
        authors
        characters {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        released
      }
      url
      name
      gender
      culture
      born
      died
      titles
      aliases
      father {
        id
        book {
          id
          url
          name
          authors
          characters
          released
        }
        url
        name
        gender
        culture
        born
        died
        titles
        aliases
        father {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        mother {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        spouse {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        allegiances
        books
        playedBy
      }
      mother {
        id
        book {
          id
          url
          name
          authors
          characters
          released
        }
        url
        name
        gender
        culture
        born
        died
        titles
        aliases
        father {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        mother {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        spouse {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        allegiances
        books
        playedBy
      }
      spouse {
        id
        book {
          id
          url
          name
          authors
          characters
          released
        }
        url
        name
        gender
        culture
        born
        died
        titles
        aliases
        father {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        mother {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        spouse {
          id
          book
          url
          name
          gender
          culture
          born
          died
          titles
          aliases
          father
          mother
          spouse
          allegiances
          books
          playedBy
        }
        allegiances
        books
        playedBy
      }
      allegiances
      books
      playedBy
    }
  }
}
"
`;

exports[`generateQuery query generates expected query with specified tab size 1`] = `
"{
     GOTHouses(name: "string") {
          id
          url
          name
          region
          titles
          seats
          words
          overlord
          swornMembers
     }
}
"
`;

exports[`generateQuery query generates expected query with string argument 1`] = `
"{
  withString(name: "string") {
    id
    url
    name
    authors
    characters
    released
  }
}
"
`;
